cranelift/codegen/src/isa/s390x/abi.rs

```diff
@@ -314,13 +315,28 @@ impl ABIMachineSpec for S390xMachineDeps {
                 } else {
                     0
                 };
-                ret.push(ABIArg::stack(
-                    (next_stack + offset) as i64,
-                    param.value_type,
-                    param.extension,
-                    param.purpose,
-                ));
+                let offset = (next_stack + offset) as i64;
                 next_stack += slot_size;
+                ABIArgSlot::Stack {
+                    offset,
+                    ty: param.value_type,
+                    extension: param.extension,
+                }
+            };
+
+            if let ir::ArgumentPurpose::StructArgument(size) = param.purpose {
+                assert!(size % 8 == 0, "StructArgument size is not properly aligned");
+                ret.push(ABIArg::StructArg {
+                    pointer: Some(slot),
+                    offset: 0,
+                    size: size as u64,
+                    purpose: param.purpose,
+                });
+            } else {
+                ret.push(ABIArg::Slots {
+                    slots: vec![slot],
```

Comment:
This will either need a rebase over @fitzgen's conversion of `slots` to `smallvec` or, if I'm misremembering what became a `smallvec`, this should become a `smallvec` :-)

Reply:
Yes, had to become a `smallvec`.   Rebased now, thanks!